L67-cpp-callback-api.md

```diff
@@ -0,0 +1,105 @@
+   - `StartRead`, `StartWrite`, `StartWritesDone`: streaming operations as appropriate to the RPC variety
+   - `StartSendInitialMetadata`: for server responses where the application wants to send the metadata separately from the first response message (uncommonly used)
+   - `Finish`: for server RPCs to provide completion status. This initiates the asynchronous transmission of the status to the client.
+1. Operation-completion reaction methods: notification of completion of asynchronous RPC activity. These are all virtual methods that default to an empty function (i.e., `{}`) but may be overridden by the application's reactor definition.
+   - `OnReadDone`, `OnWriteDone`, `OnWritesDoneDone`: methods invoked by the library at the completion of an explicitly-specified streaming RPC operation. Provides a `bool ok` argument to indicate whether the completion was successful.
+   - `OnReadInitialMetadataDone`: notifying client RPCs that the server has sent initial metadata. It has a bool `ok` argument to indicate whether the completion was successful. This is automatically invoked by the library for all RPCs; it is uncommonly used however
+   - `OnCancel`: for server RPCs only, called if an RPC is canceled before it has a chance to successfully send status to the client side
+   - `OnDone`: called when all outstanding RPC operations are completed. Provides status on the client side. 
+1. RPC completion prevention methods. These are methods provided by the class and are not virtual.
+   - `AddHold`, `RemoveHold`: for client RPCs, prevents the RPC from being considered complete (ready for `OnDone`) until each AddHold is matched to a corresponding RemoveHold. These are used to perform _extra-reaction flows_, which refers to  streaming operations initiated from outside a reaction method or method handler.
+
+### Unary RPC shortcuts
+
+As a shortcut, client-side unary RPCs _may_ bypass the reactor model by directly providing a `std::function` for the library to call at completion rather than a reactor object pointer. This is semantically equivalent to a reactor in which the `OnDone` function simply invokes the specified function (but can be implemented in a slightly faster way since such an RPC will definitely not wait separately for initial metadata from the server). In practice, this is the common and recommended model for client-side unary RPCs, unless they have a specific need to wait for initial metadata before getting their full response message.
+
+Server-side unary RPCs have the option of returning a library-provided default reactor when there method handler is invoked. This default reactor provides a Finish method, but does not provide a user callback for `OnCancel` and `OnDone`. In practice, this is the common and recommended model for most server-side unary RPCs unless they specifically need to react to an `OnCancel` callback or do cleanup work after the RPC fully completes.
```

Comment:
Where are the method handlers methods run? Can a slow handler method prevent the rest of the system from making forward progress? Are there any mitigations for this?

My experience implementing such callback APIs has been that user callbacks, by default, need to run on separate threads from the network/IO threads. This helps mitigate deadlocks when handlers inadvertently do blocking work. An "I solemnly swear I won't block" escape hatch can be provided for apps that need the performance gained by one fewer context switch. Another approach I've seen is to periodically inject another worker thread when they're all block to see if that frees things up.

Reply:
Absolutely correct. I added it in the comments and will be adding it to the proposal.